jobs/build-charms/charms.py

```diff
@@ -157,40 +160,83 @@ def save(self):
         self.db_json.write_text(json.dumps(dict(self.db)))
         self.store.put_item(Item=dict(self.db))
 
-    def promote_all(self, from_channel="unpublished", to_channel="edge"):
+    def _cs_promote(self, charm_namespace, charm_name, from_channel, to_channel):
+        charm_entity = f"cs:~{charm_namespace}/{charm_name}"
+        click.echo(
+            f"Promoting :: {charm_entity:^35} :: from:{from_channel} to: {to_channel}"
+        )
+        charm_id = sh.charm.show(charm_entity, "--channel", from_channel, "id")
+        charm_id = yaml.safe_load(charm_id.stdout.decode())
+        resources_args = []
+        try:
+            resources = sh.charm(
+                "list-resources",
+                charm_id["id"]["Id"],
+                channel=from_channel,
+                format="yaml",
+            )
+            resources = yaml.safe_load(resources.stdout.decode())
+            if resources:
+                resources_args = [
+                    (
+                        "--resource",
+                        "{}-{}".format(resource["name"], resource["revision"]),
+                    )
+                    for resource in resources
+                ]
+        except sh.ErrorReturnCode_1:
+            click.echo("No resources for {}".format(charm_id))
+        sh.charm.release(
+            charm_id["id"]["Id"], "--channel", to_channel, *resources_args
+        )
+
+    def _charmcraft_status(self, charm_entity):
+        """Read CLI Table output from charmcraft status and parse"""
+        charm_status = sh.charmcraft.status(charm_entity)
+        header, *body = charm_status.stdout.decode().splitlines()
+        channel_status = []
+        for line in body:
+            row, head = {}, line
+            for key in reversed(header.split()):
+                head, *value = head.strip().rsplit("  ", 1)
+                head, value = ("", head.strip()) if not value else (head, value[0])
+                row[key] = value or channel_status[-1].get(key)
+            resources = row.get('Resources', '')
+            if resources == '-':
+                row['Resources'] = []
+            elif resources == '↑':
+                row['Resources'] = channel_status[-1]['Resources']
+            else:
+                row['Resources'] = self.CHARMCRAFT_RESOURCE.findall(resources)
+            channel_status.append(row)
+        return channel_status
```

Comment:
helper to parse output of `charmcraft status <charm>` because it doesn't provide a yaml format output